packages/ember-application/lib/system/application-instance.js

```diff
@@ -212,6 +298,255 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
   }
 });
 
+if (isEnabled('ember-application-visit')) {
+  ApplicationInstance.reopen({
+    /**
+      Returns the current URL of the app instance. This is useful when your
+      app does not update the browsers URL bar (i.e. it uses the `'none'`
+      location adapter).
+
+      @public
+      @return {String} the current URL
+    */
+    getURL() {
+      let router = get(this, 'router');
+      return get(router, 'url');
+    },
+
+    // `instance.visit(url)` should eventually replace `instance.handleURL()`;
+    // the test helpers can probably be switched to use this implementation too
+
+    /**
+      Navigate the instance to a particular URL. This is useful in tests, for
+      example, or to tell the app to start at a particular URL. This method
+      returns a promise that resolves with the app instance when the transition
+      is complete, or rejects if the transion was aborted due to an error.
+
+      @public
+      @param url {String} the destination URL
+      @return {Promise}
+    */
+    visit(url) {
+      this.setupRouter();
+
+      let router = get(this, 'router');
+
+      let handleResolve = () => {
+        // Resolve only after rendering is complete
+        return new Ember.RSVP.Promise((resolve) => {
+          // TODO: why is this necessary? Shouldn't 'actions' queue be enough?
+          // Also, aren't proimses supposed to be async anyway?
+          run.next(null, resolve, this);
+        });
+      };
+
+      let handleReject = (error) => {
+        if (error.error) {
+          throw error.error;
+        } else if (error.name === 'TransitionAborted' && router.router.activeTransition) {
+          return router.router.activeTransition.then(handleResolve, handleReject);
+        } else if (error.name === 'TransitionAborted') {
+          throw new Error(error.message);
+        } else {
+          throw error;
+        }
```

Comment:
I'm not 100% sure this is correct, someone pointed out to me that using `router.router.activeTransition` might be (incorrectly) relying on sync behavior. (Although I see other places using it too.)

However, since the docs/tests is describing the intended behavior (it should follow all redirects and then resolve when rending work is complete, if any), and this implementation is correct as far as the current tests are concerned, if no one has any immediate suggestions this shouldn't be a blocker either. If someone eventually discovered the theoretical bug we can change the implementation to match the intended semantics.